src/prefect/runner/storage.py

```diff
@@ -0,0 +1,206 @@
+from pathlib import Path
+from typing import Optional, Protocol, runtime_checkable
+from urllib.parse import urlparse, urlunparse
+
+from anyio import run_process
+
+from prefect.logging.loggers import get_logger
+
+
+@runtime_checkable
+class RunnerStorage(Protocol):
+    """
+    A storage interface for a runner to use to retrieve
+    remotely stored flow code.
+    """
+
+    def set_base_path(self, path: Path):
+        """
+        Sets the base path to use when syncing remote storage to local storage.
+        """
+        ...
+
+    @property
+    def sync_interval(self) -> Optional[int]:
+        """
+        The interval at which remote storage should be synced to local storage.
+        If None, remote storage will perform a one-time sync.
+        """
+        ...
+
+    @property
+    def destination(self) -> Path:
+        """
+        The local file path to sync remote storage to.
+        """
+        ...
+
+    async def sync(self):
```

Comment:
I want to propose calling this `pull` (or some other unidirectional verb) so that we can incorporate pushing results to remote locations on runners as well.

Reply:
Yeah, `sync` does imply that changes to the local filesystem could affect the remote filesystem, which would be inaccurate. `pull` also feels more consistent with the existing pull steps. I'll make that change.